src/journal/bsod.c

```diff
@@ -0,0 +1,143 @@
+/* SPDX-License-Identifier: LPGL-2.1-or-later */
+
+#include <errno.h>
+#include <getopt.h>
+#include <stdio.h>
+
+#include "alloc-util.h"
+#include "build.h"
+#include "fd-util.h"
+#include "log.h"
+#include "main-func.h"
+#include "sd-id128.h"
+#include "sd-journal.h"
+#include "terminal-util.h"
+
+static void help(void) {
+        printf("%s\n\n"
+               "filters the journal to fetch the first message from the current boot with an emergency log level.\n\n"
+               "   -h --help            Show this help\n"
+               "      --version         Show package version\n",
+               program_invocation_short_name);
+}
+
+static char* first_emerg_boot_message(void) {
+        sd_journal *j;
+        const void *d;
+        size_t l;
+        char boot_id_string[33];
+        sd_id128_t boot_id;
+        char boot_id_filter[42];
+        int r;
+        _cleanup_free_ char *message = NULL;
+
+        r = sd_journal_open(&j, SD_JOURNAL_LOCAL_ONLY);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to open journal: %m");
+                return NULL;
+        }
+
+        r = sd_id128_get_boot(&boot_id);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to get boot ID, ignoring : %m");
+                goto clean_journal;
+        }
+        sd_id128_to_string(boot_id, boot_id_string);
+
+        snprintf(boot_id_filter, sizeof(boot_id_filter), "_BOOT_ID=%s", boot_id_string);
+        r = sd_journal_add_match(j, boot_id_filter, 0);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to add boot ID filter: %m");
+                goto clean_journal;
+        }
+
+        r = sd_journal_add_match(j, "_UID=0", 0);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to add User ID filter: %m");
+                goto clean_journal;
+        }
+
+        r = sd_journal_add_match(j, "PRIORITY=0", 0);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to add Emergency filter: %m");
+                goto clean_journal;
+        }
+
+        SD_JOURNAL_FOREACH(j) {
```

Comment:
don't use a loop if you just want the first message. Use sd_journal_seek_first() once, plus one sd_journal_next() instead